### PR TITLE
Add separator version of DT_FOREACH* macros

### PR DESCRIPTION
**Pull request by gmarull**
It is frequent to see in Devicetree code constructs like:

```c
#define LABEL_AND_COMMA(node_id) DT_LABEL(node_id),

const char *child_labels[] = {
	DT_FOREACH_CHILD(DT_NODELABEL(n), LABEL_AND_COMMA)
};
```

That is, an auxiliary macro to append a separator character in
DT_FOREACH* macros. Non-DT API, e.g. FOR_EACH(), takes a separator
argument to avoid such intermediate macros.

This patch adds DT_FOREACH_CHILD_SEP (and instance version
of it). It takes an extra argument: a separator. With this change, the
example above can be simplified to:

```c
const char *child_labels[] = {
	DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,))
};
```

Notes:
- Other DT_FOREACH* macros could/should be extended as well
- We could modify original macros to always take a separator, to align
  with non-DT FOREACH* macros, but that would be a breaking change
- Placement of `sep` argument: `DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,))` or `DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), (,), DT_LABEL)`? (relevant for the VARGS case)

### CODE REVIEW

**Comment by galak** [comment]
I think its best to keep the original macro as is and add a new macro.

Do we need a version for `DT_FOREACH_CHILD_VARGS`?

**Comment by gmarull** [comment]
> I think its best to keep the original macro as is and add a new macro.
> 
> Do we need a version for `DT_FOREACH_CHILD_VARGS`?

Yes, definitely. `DT_FOREACH_PROP_ELEM` probably as well.

**Comment by gmarull** [comment]
> Do we have any other in-tree usages? My only concern is how often is this pattern used to introduce a new API?

Grepping for `AND_COMMA` shows a few, there are others with other namings, e.g. in `subsys/pm/state.c`.

**Comment by gmarull** [comment]
> You asked about argument order offline -- I don't see any specific questions on that in the thread here but it looks generally OK to me.

It's in the PR description:

> Placement of sep argument: DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,)) or DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), (,), DT_LABEL)? (relevant for the VARGS case)

The reason is that with current approach, separator will go between function and VARGS, which may be confusing. If it goes before, we'll have separator + function + vargs.

**Comment by mbolivar-nordic** [comment]
I personally think this order looks good:

```
DT_FOREACH_FOO_SEP(node_id, fn, sep)
DT_FOREACH_FOO_SEP_VARGS(node_id, fn, sep, ...)
```

I'm not married to it, though

File: include/zephyr/devicetree.h [hunk 1]
```diff
@@ -1927,6 +1927,43 @@
...
+ *             };
+ *             child-2 {
+ *                     label = "bar";
+ *             };
+ *     };
+ *
+ * Example usage:
+ *
+ *     const char *child_labels[] = {
+ *         DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,))
```
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:1950]: Let's not use DT_LABEL given the removal of those properties. How about DT_NODE_FULL_NAME instead?

File: subsys/pm/state.c [hunk 2]
```diff
@@ -39,23 +39,21 @@ BUILD_ASSERT(DT_NODE_EXISTS(DT_PATH(cpus)),
 /* Check that all power states are consistent */
 DT_FOREACH_CHILD(DT_PATH(cpus), CHECK_POWER_STATES_CONSISTENCY)
 
-#define NUM_CPU_STATES(n) DT_NUM_CPU_POWER_STATES(n),
```
Comment (mbolivar-nordic) [inline subsys/pm/state.c:42]: Looks like DT_NUM_CPU_POWER_STATES was added by you in 7bfd0976aad, with a potential move to devicetree.h in the future. Gentle reminder that this is still an abuse of the DT_ namespace.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: subsys/pm/state.c (hunk 2 above)
Comment (gmarull) [reply subsys/pm/state.c:42]: thanks for the reminder here, will likely move that macro to DT API

File: scripts/dts/gen_defines.py [hunk 3]
```diff
@@ -267,6 +269,13 @@ def write_top_comment(edt):
     out_comment(s, blank_before=False)
 
 
+def write_utils():
+    # Writes utility macros
+
+    out_comment("Used to remove brackets from around a single argument")
+    out_define("__DT_DEBRACKET(...)", "__VA_ARGS__")
```
[the remark this answers is not in the document]
Comment (gmarull) [reply scripts/dts/gen_defines.py]: done

File: include/zephyr/devicetree.h (hunk 1 above)
Comment (gmarull) [reply include/zephyr/devicetree.h]: updated

**Review by mbolivar-nordic — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**